### ISSUE TITLE
Could not find melpa recipe even though its file exists

### ISSUE THREAD

**Issue by publicimageltd** [raxod502/straight.el#835]
First, thanks for `straight.el`. After hesitating quite a while, I have finally switched to it. I like it; it feels fast, safe and well thought-out. 

Today, I wanted to install the package `all-the-icons-completion`. So I added the following snippet to my `.init`:

```emacs-lisp
(use-package all-the-icons-completion
  :after all-the-icons
  :config
  (all-the-icons-completion-mode))
```

To my surprise, `straight` tells me that there is a malformed recipe. Debugging `straight-get-recipe`, I see that it does not find the associated recipe. Looking at `.emacs.d/straight/repos/melpa/recipes/all-the-icons-completion`, however, I find the recipe all right. I even edited its formatting, removing whitespace, just to check. No change. Issuing `straight-pull-recipe-repositories` changed nothing, either (besides a warning that I had changed the recipe file, very well done!). 

I now added my hand-made recipe:

```emacs-lisp
:straight (all-the-icons-completion
             :repo "iyefrat/all-the-icons-completion"
             :host github
             :type git)
```

This works, but why does the normal melpa-way not work? All the other packages which use melpa work out of the box (or "straight away", if I might say). Any ideas? I don't even know how to debug it.

**Comment by progfolio** [comment]
> First, thanks for straight.el. After hesitating quite a while, I have finally switched to it. I like it; it feels fast, safe and well thought-out.

Thank you for the kind words. We're glad you're finding it useful.

> Today, I wanted to install the package all-the-icons-completion. So I added the following snippet to my .init:
> 
> (use-package all-the-icons-completion
>   :after all-the-icons
>   :config
>   (all-the-icons-completion-mode))
> 
> To my surprise, straight tells me that there is a malformed recipe.
> Debugging straight-get-recipe, I see that it does not find the associated recipe.
> Looking at .emacs.d/straight/repos/melpa/recipes/all-the-icons-completion, however, I find the recipe all right.
> I even edited its formatting, removing whitespace, just to check. No change.

What you describe here sounds like the MELPA recipe repository needing to be pulled, though if the recipe file is present it should have worked.

> Issuing straight-pull-recipe-repositories changed nothing, either (besides a warning that I had changed the recipe file, very well done!).

`straight-pull-recipe-repositories` or `M-x straight-pull-package melpa` should both do the trick in this case, but only if the repo needs to be updated.

> I now added my hand-made recipe:
> 
> :straight (all-the-icons-completion
>              :repo "iyefrat/all-the-icons-completion"
>              :host github
>              :type git)
> 
> This works,&#x2026;

Glad you found a workaround for the time being.

> but why does the normal melpa-way not work?
> &#x2026; Any ideas? I don't even know how to debug it.

It's hard to say without being able to reproduce it on my end, but we have a tool to debug called `straight-bug-report`.
You can evaluate the following test case in your \*scratch\* buffer.
It will install straight.el in a clean, temporary environment and attempt to install all-the-icons-completion.
When the test has completed, a report buffer should pop up in your Emacs session, which you can copy into this issue.


<details><summary>Test Case</summary>

```emacs-lisp
(straight-bug-report
  :post-bootstrap 
  (straight-use-package 'all-the-icons-completion))
```
</details>

- Test run at: `2021-08-12 13:20:41`
- system-type: `gnu/linux`
- straight-version: `prerelease (HEAD -> develop, origin/develop) 08b0ecf 2021-07-24`
- emacs-version: `GNU Emacs 28.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.30, cairo version 1.17.4)
 of 2021-08-02`

<details><summary>Output</summary>

```emacs-lisp
Bootstrapping straight.el...
Bootstrapping straight.el...done
Rebuilding all packages due to build cache schema change
Looking for gnu-elpa-mirror recipe -> Cloning melpa...
Looking for gnu-elpa-mirror recipe -> Cloning melpa...done
Looking for emacsmirror-mirror recipe -> Cloning gnu-elpa-mirror...
Looking for emacsmirror-mirror recipe -> Cloning gnu-elpa-mirror...done
Looking for emacsmirror-mirror recipe -> Cloning el-get...
Looking for emacsmirror-mirror recipe -> Cloning el-get...done
Looking for straight recipe -> Cloning emacsmirror-mirror...
Looking for straight recipe -> Cloning emacsmirror-mirror...done
Building straight...
Building straight...done

Test run with version: prerelease (HEAD -> develop, origin/develop) 08b0ecf 2021-07-24
Cloning all-the-icons-completion...
Cloning all-the-icons-completion...done
Building all-the-icons-completion...
Building all-the-icons-completion -> Cloning all-the-icons.el...
Building all-the-icons-completion -> Cloning all-the-icons.el...done
Building all-the-icons-completion -> Building all-the-icons...
Building all-the-icons-completion -> Building all-the-icons...done
Building all-the-icons-completion...
Building all-the-icons-completion...done

Packages:
"org-elpa"                n/a                  n/a
"melpa"                   n/a                  master c44eceaf 2021-08-09
"gnu-elpa-mirror"         n/a                  master 8ce05c5 2021-08-10
"el-get"                  melpa                master 463f5e98 2021-07-06
"emacsmirror-mirror"      n/a                  master 664c0cc 2021-08-09
"straight"                n/a                  develop 08b0ecf 2021-07-24
"all-the-icons-completion" melpa                master 9650041 2021-08-08
"all-the-icons"           melpa                master e01eeb4 2021-07-29


```
</details>

**Comment by publicimageltd** [comment]
Wow, that's an insanely easy way to test if the bug is reproducible in a clean environment!

Unfortunately, unless you can see anything problematic in the output pasted below, it looks like the problem is on my side. Would "normalizing" the packages be a good idea? I had frozen them after my first fresh install so to be able to return the first installation version; should I somehow delete that version lockfile before normalizing it? I admit this part of straight is not yet clear to me.


<details><summary>Test Case</summary>

```emacs-lisp
(straight-bug-report
  :post-bootstrap 
  (straight-use-package 'all-the-icons-completion))
```
</details>

- Test run at: `2021-08-12 19:51:44`
- system-type: `gnu/linux`
- straight-version: `prerelease (HEAD -> master, origin/master, origin/HEAD) 1e27b05 2021-06-13`
- emacs-version: `GNU Emacs 27.2 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.27, cairo version 1.17.4)
 of 2021-03-26`

<details><summary>Output</summary>

```emacs-lisp
Bootstrapping straight.el...
Bootstrapping straight.el...done
Rebuilding all packages due to build cache schema change
Looking for gnu-elpa-mirror recipe → Cloning melpa...
Looking for gnu-elpa-mirror recipe → Cloning melpa...done
Looking for emacsmirror-mirror recipe → Cloning gnu-elpa-mirror...
Looking for emacsmirror-mirror recipe → Cloning gnu-elpa-mirror...done
Looking for emacsmirror-mirror recipe → Cloning el-get...
Looking for emacsmirror-mirror recipe → Cloning el-get...done
Looking for straight recipe → Cloning emacsmirror-mirror...
Looking for straight recipe → Cloning emacsmirror-mirror...done
Building straight...
Building straight...done

Test run with version: prerelease (HEAD -> develop, origin/develop) 08b0ecf 2021-07-24
Cloning all-the-icons-completion...
Cloning all-the-icons-completion...done
Building all-the-icons-completion...
Building all-the-icons-completion → Cloning all-the-icons.el...
Building all-the-icons-completion → Cloning all-the-icons.el...done
Building all-the-icons-completion → Building all-the-icons...
Building all-the-icons-completion → Building all-the-icons...done
Building all-the-icons-completion...
Building all-the-icons-completion...done

Packages:
"org-elpa"                n/a                  n/a
"melpa"                   n/a                  master c44eceaf 2021-08-09
"gnu-elpa-mirror"         n/a                  master 8ce05c5 2021-08-10
"el-get"                  melpa                master 463f5e98 2021-07-06
"emacsmirror-mirror"      n/a                  master 664c0cc 2021-08-09
"straight"                n/a                  develop 08b0ecf 2021-07-24
"all-the-icons-completion" melpa                master 9650041 2021-08-08
"all-the-icons"           melpa                master e01eeb4 2021-07-29


```
</details>

**Comment by progfolio** [comment]
"Public Image Ltd." ***@***.***> writes:

> Wow, that's an insanely easy way to test if the bug is 
> reproducible in a clean environment!

Thanks, it's been valuable in finding bugs and developing 
features.
 
> Unfortunately, unless you can see anything problematic in the 
> output pasted below, it looks like the
> problem is on my side.

The output looks like everything installed properly.

> Would "normalizing" the packages be a good idea? I had frozen 
> them after
> my first fresh install so to be able to return the first 
> installation version; should I somehow delete
> that version lockfile before normalizing it? I admit this part 
> of straight is not yet clear to me.

Ah, I wasn't aware you were using the lockfile system as well.
I believe there are some corner cases with freezing/thawing that 
still need to be worked out.
See #292, which may be relevant.

In general, `straight-freeze-versions` will write the current 
version for each package to the lockfile.
`straight-thaw-versions` will check out each package to the 
version described in your lockfile.
So you're defining a stable set of packages when freezing and 
returning your packages to that state when thawing.

Normalizing a package can be thought of a way to put a package's 
local repository in a state that matches its recipe.
e.g.

```emacs-lisp
;;This call registers the recipe for the `example` package.
(straight-use-package '(example :branch "main"))

;;Straight will issue a warning here, because it detects that 
  there are now two different recipes for the `example` package:
(straight-use-package '(example :branch "develop"))
;; Warning (straight): Two different recipes given for "example" 
   (:branch cannot be both "main" and "develop")
```

Running `M-x straight-normalize-package example` at this point 
would attempt to put example's local repository on the "develop" 
branch. Likewise if any other aspect of the recipe changed, it 
would attempt to rectify that.

Does that help?

**Comment by publicimageltd** [comment]
Thanks for the hint to the issue. I "thawed" everything, then I ultimately just removed the version file on disk, and called "normalize-all". This found some weird errors (local packages sometimes issued an error), which I might try to reproduce some time in the future and post as an issue if they persist. But I got caught in the weird problem that even I tried to force rebuilding of a local repo, it was not built. I then gave it up, but restarting emacs did the trick. So in short: I hardly understand what's going on, I will need to have a closer look at the documentation, and until then, my workaround for the original problem posted here does its due.

Maybe this setting is part of the problem:
```emacs-lisp
(setq straight-check-for-modifications nil)
```

I have this set so that I have full control over when new versions are pulled and built. Maybe that's a reason why the recipe isn't found? IDK. And after two hours playing with Emacs configurations, I tend to let it as it is, so I close the issue.
